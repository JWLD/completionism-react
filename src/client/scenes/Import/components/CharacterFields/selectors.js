import { createSelector } from 'reselect'
import { formValueSelector } from 'redux-form'

const importFormSelector = formValueSelector('import')

const realmListSelector = state => state.import.realms

export const regionSelector = state =>
  importFormSelector(state, 'character.region')

const regionRealmsSelector = createSelector(
  [realmListSelector, regionSelector],
  (realmList, region) => realmList[region] || []
)

export const realmDataSelector = createSelector(
  [regionRealmsSelector],
  realms =>
    realms.reduce((realmList, realm) => {
      realmList[realm.slug] = realm.name

      return realmList
    }, {})
)
