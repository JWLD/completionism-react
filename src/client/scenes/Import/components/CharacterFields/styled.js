import styled from 'styled-components'

import { mixins } from 'style'

const inputFont = '2.4rem DinMC'
const inputHeight = 4
const inputSpacing = 1
const indicatorSize = inputHeight * 2 + inputSpacing

export const CharacterFieldsWrap = styled.div`
  ${mixins.flex};

  flex-wrap: wrap;
  width: 100%;
`

export const RegionSelect = styled.select`
  height: ${inputHeight}rem;
  width: 10rem;
  margin-right: ${inputSpacing}rem;
  font: ${inputFont};
`

export const RealmSelect = RegionSelect.extend`
  flex-grow: 1;
  margin-right: 0;
`

export const NameInput = styled.input`
  width: 100%;
  height: ${inputHeight}rem;
  margin-top: ${inputSpacing}rem;
  padding: 0.8rem;
  border-radius: 0.4rem;
  font: ${inputFont};
  text-transform: capitalize;
`
