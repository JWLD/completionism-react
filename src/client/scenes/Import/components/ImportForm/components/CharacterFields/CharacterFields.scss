@import 'global';

$inputHeight: 4rem;
$boxPadding: 2rem;

.char-sctn {
	.char-inputs {
		flex-wrap: wrap;
		flex-grow: 1;

		select, input {
			height: $inputHeight;

			border-radius: .3rem;
			font: 2.5rem DinMC;
		}

		select {
			&:first-child {
				width: 10rem;
				margin-right: $boxPadding * .5;
			}

			&:nth-child(2) {
				flex-grow: 1;
			}
		}

		input {
			width: 100%;
			margin-top: $boxPadding * .5;
			padding: .8rem;

			text-transform: capitalize;
		}
	}

	button {
		height: ($inputHeight * 2) + ($boxPadding * .5);
		min-width: ($inputHeight * 2) + ($boxPadding * .5);
		margin-left: $boxPadding;
		padding: 2rem;

		border-radius: .4rem;
		background-color: $cmpGrey;

		svg {
			height: 100%;
			width: 100%;
		}

		&:hover {
			background-color: #555;
		}
	}

	.import-err {
		margin-top: $pagePadding * .5;

		color: $neg;
		font: 2.5rem DinRC;
	}
}
