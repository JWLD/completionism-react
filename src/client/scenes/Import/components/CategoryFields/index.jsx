import React from 'react'
import { Field } from 'redux-form'

import { CATEGORIES_BY_SECTION as CATEGORIES } from 'constants/categories'

import * as SC from 'CategoryFields/styled'
import { CheckboxField } from 'formFields'

const CategoryFields = () => {
  const renderInputRows = categories => {
    return categories.map(({ battleNet, key, name }) => {
      if (!battleNet) return null

      return (
        <SC.CheckboxLabel key={key}>
          {name}
          <Field
            component={CheckboxField}
            name={key}
            StyledComponent={SC.CheckboxInput}
          />
          <SC.FakeInput className="fa fa-check" />
        </SC.CheckboxLabel>
      )
    })
  }

  const renderCheckboxColumns = () => {
    return Object.keys(CATEGORIES).map(category => (
      <SC.CheckboxWrap key={category}>
        {renderInputRows(CATEGORIES[category])}
      </SC.CheckboxWrap>
    ))
  }

  return renderCheckboxColumns()
}

export default CategoryFields
