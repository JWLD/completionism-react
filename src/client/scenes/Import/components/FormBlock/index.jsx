import React from 'react'
import PropTypes from 'prop-types'
import { FormSection } from 'redux-form'

import * as SC from 'FormBlock/styled'

const FormBlock = props => (
  <SC.FormSection>
    <SC.FormHeader>{props.header}</SC.FormHeader>

    <SC.InputWrap>
      <FormSection name={props.name}>{props.fieldsComponent}</FormSection>
    </SC.InputWrap>
  </SC.FormSection>
)

FormBlock.propTypes = {
  fieldsComponent: PropTypes.object.isRequired,
  header: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired
}

export default FormBlock
