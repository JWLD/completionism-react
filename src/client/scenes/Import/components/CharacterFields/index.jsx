import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Field } from 'redux-form'

import { realmDataSelector, regionSelector } from './selectors'
import { REGIONS } from 'constants/blizzard'
import { fetchRealmData } from 'scenes/Import/actions'

import * as SC from './styled'
import { InputField, SelectBoxField } from 'components/ReduxFields'

class CharacterFields extends Component {
  componentDidMount() {
    this.getRealmData(this.props.region)
  }

  onRegionChange = (changeEvent) => {
    const newRegion = changeEvent.target.value

    this.getRealmData(newRegion)
  }

  getRealmData(region) {
    if (!this.props.realmList[region]) {
      this.props.fetchRealmData(region)
    }
  }

  render() {
    return (
      <SC.CharacterFieldsWrap>
        <Field
          component={SelectBoxField}
          name="region"
          onChange={this.onRegionChange}
          options={REGIONS}
          StyledComponent={SC.RegionSelect}
        />

        <Field
          component={SelectBoxField}
          name="realm"
          options={this.props.realmData}
          placeholder="Select Realm"
          StyledComponent={SC.RealmSelect}
        />

        <Field component={InputField} name="name" placeholder="Name" StyledComponent={SC.NameInput} />
      </SC.CharacterFieldsWrap>
    )
  }
}

CharacterFields.propTypes = {
  fetchRealmData: PropTypes.func.isRequired,
  realmData: PropTypes.object.isRequired,
  realmList: PropTypes.object.isRequired,
  region: PropTypes.string.isRequired
}

const mapStateToProps = state => ({
  realmData: realmDataSelector(state),
  realmList: state.import.realms,
  region: regionSelector(state)
})

const mapDispatchToProps = {
  fetchRealmData
}

export default connect(mapStateToProps, mapDispatchToProps)(CharacterFields)
