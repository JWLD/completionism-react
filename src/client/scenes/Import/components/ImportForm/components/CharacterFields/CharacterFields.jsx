import React from 'react';

import './CharacterFields.scss';

const CharacterFields = () => (
	<section className="sctn char-sctn">
		<h2 className="sctn__header">Select Character</h2>

		<div className="sctn__input-wrap">
			<div className="char-inputs">
				<select>
					<option value="eu">EU</option>
					<option value="us">US</option>
				</select>

				<select>
					{/* {realms} */}
				</select>

				<input placeholder="Name" />
			</div>

			<button>
				{/* {symbol} */}
			</button>
		</div>

		{/* {errBox} */}
	</section>
);

export default CharacterFields;
