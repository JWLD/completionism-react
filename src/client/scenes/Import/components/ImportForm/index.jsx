import React from 'react'
import PropTypes from 'prop-types'
import { reduxForm } from 'redux-form'

import { fetchCharData } from './utils'

import FormBlock from 'FormBlock'
import CategoryFields from 'CategoryFields'
import CharacterFields from 'CharacterFields'

export const ImportForm = props => (
  <form onSubmit={props.handleSubmit(fetchCharData)}>
    <FormBlock
      header="Select Categories"
      name="categories"
      fieldsComponent={<CategoryFields />}
    />
    <FormBlock
      header="Select Character"
      name="character"
      fieldsComponent={<CharacterFields />}
    />
  </form>
)

ImportForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
}

const FormWrapper = reduxForm({
  form: 'import',
  initialValues: {
    character: {
      region: 'eu'
    }
  }
})

export default FormWrapper(ImportForm)
