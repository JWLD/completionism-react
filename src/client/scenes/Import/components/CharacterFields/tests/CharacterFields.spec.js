import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { CharacterFields } from 'scenes/Import/components/CharacterFields'

Enzyme.configure({ adapter: new Adapter() })

const mockProps = {
  fetchRealmData: jest.fn(),
  realmData: {},
  realmList: {},
  region: ''
}

describe('<CharacterFields />', () => {
  test('renders correctly', () => {
    const component = shallow(<CharacterFields {...mockProps} />)

    expect(component).toMatchSnapshot()
  })
})
