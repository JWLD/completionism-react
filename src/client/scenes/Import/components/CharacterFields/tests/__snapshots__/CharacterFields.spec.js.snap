// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterFields /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterFields
    fetchRealmData={
      [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
      }
    }
    realmData={Object {}}
    realmList={Object {}}
    region=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Field
          StyledComponent={[Function]}
          component={[Function]}
          name="region"
          onChange={[Function]}
          options={
            Object {
              "eu": "EU",
              "us": "US",
            }
          }
        />,
        <Field
          StyledComponent={[Function]}
          component={[Function]}
          name="realm"
          options={Object {}}
          placeholder="Select Realm"
        />,
        <Field
          StyledComponent={[Function]}
          component={[Function]}
          name="name"
          placeholder="Name"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "StyledComponent": [Function],
          "component": [Function],
          "name": "region",
          "onChange": [Function],
          "options": Object {
            "eu": "EU",
            "us": "US",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "StyledComponent": [Function],
          "component": [Function],
          "name": "realm",
          "options": Object {},
          "placeholder": "Select Realm",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "StyledComponent": [Function],
          "component": [Function],
          "name": "name",
          "placeholder": "Name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Field
            StyledComponent={[Function]}
            component={[Function]}
            name="region"
            onChange={[Function]}
            options={
              Object {
                "eu": "EU",
                "us": "US",
              }
            }
          />,
          <Field
            StyledComponent={[Function]}
            component={[Function]}
            name="realm"
            options={Object {}}
            placeholder="Select Realm"
          />,
          <Field
            StyledComponent={[Function]}
            component={[Function]}
            name="name"
            placeholder="Name"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "StyledComponent": [Function],
            "component": [Function],
            "name": "region",
            "onChange": [Function],
            "options": Object {
              "eu": "EU",
              "us": "US",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "StyledComponent": [Function],
            "component": [Function],
            "name": "realm",
            "options": Object {},
            "placeholder": "Select Realm",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "StyledComponent": [Function],
            "component": [Function],
            "name": "name",
            "placeholder": "Name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
